src/main/java/io/cdap/plugin/google/common/GoogleDriveBaseConfig.java

```diff
@@ -67,49 +102,80 @@
   public void validate(FailureCollector collector) {
     IdUtils.validateReferenceName(referenceName, collector);
 
-    if (validateAccountFilePath(collector)) {
-      try {
-        GoogleDriveClient client = getDriveClient();
-
-        // validate auth
-        validateCredentials(collector, client);
+    if (validateAuthType(collector)) {
+      AuthType authType = getAuthType();
+      boolean propertiesAreValid = false;
+      switch (authType) {
+        case OAUTH2:
+          propertiesAreValid = validateOAuth2Properties(collector);
+          break;
+        case SERVICE_ACCOUNT:
+          propertiesAreValid = validateAccountFilePath(collector);
+          break;
+      }
+      if (propertiesAreValid) {
+        try {
+          GoogleDriveClient client = getDriveClient();
+
+          // validate auth
+          validateCredentials(collector, client);
+
+          // validate directory
+          validateDirectoryIdentifier(collector, client);
+
+        } catch (Exception e) {
+          collector.addFailure(
+            String.format("Exception during authentication/directory properties check: %s", e.getMessage()),
+            "Check message and reconfigure the plugin")
+            .withStacktrace(e.getStackTrace());
+        }
+      }
+    }
+  }
 
-        // validate directory
-        validateDirectoryIdentifier(collector, client);
+  protected abstract GoogleDriveClient getDriveClient();
 
-      } catch (Exception e) {
-        collector.addFailure(
-          String.format("Exception during authentication/directory properties check: %s", e.getMessage()),
-          "Check message and reconfigure the plugin")
-          .withStacktrace(e.getStackTrace());
+  private boolean validateAuthType(FailureCollector collector) {
+    if (!containsMacro(AUTH_TYPE)) {
+      try {
+        getAuthType();
+        return true;
+      } catch (InvalidPropertyTypeException e) {
+        collector.addFailure(String.format("Auth type has invalid value '%s'", authType),
+                             "Provide correct auth type")
+          .withConfigProperty(AUTH_TYPE).withStacktrace(e.getStackTrace());
```

Comment:
stackTrace does not provide any additional information in this case. We should not add it.

Reply:
done